### PR TITLE
Add Spec to ServiceClass and ServicePlan

### LINKED ISSUE #1214: ServiceClass and ServicePlan should each have spec and status
I spent some time thinking about different facets of these resources this weekend.  I believe that they should have spec and status.  It's likely that we will want controllers to maintain status information for both of these resources.

Depends on #1106

### PR DESCRIPTION
**Pull request by vaikas-google**
Closes #1214

### CODE REVIEW

**Comment by vaikas-google** [comment]
This is work in progress, just sharing the Spec changes earlier for a chance of feedback in case people have concerns. Tests are still b0rk3d and I'll work through those.

File: pkg/registry/servicecatalog/serviceclass/storage.go [hunk 1]
```diff
@@ -89,8 +89,13 @@ func Match(label labels.Selector, field fields.Selector) storage.SelectionPredic
...
 // toSelectableFields returns a field set that represents the object for matching purposes.
 func toSelectableFields(serviceClass *servicecatalog.ServiceClass) fields.Set {
-	objectMetaFieldsSet := generic.ObjectMetaFieldsSet(&serviceClass.ObjectMeta, true)
-	return generic.MergeFieldsSets(objectMetaFieldsSet, nil)
+	// The purpose of allocation with a given number of elements is to reduce
+	// amount of allocations needed to create the fields.Set. If you add any
+	// field here or the number of object-meta related fields changes, this should
+	// be adjusted.
+	scSpecificFieldsSet := make(fields.Set, 1)
+	scSpecificFieldsSet["spec.externalName"] = serviceClass.Spec.ExternalName
```
Comment (pmorie) [inline pkg/registry/servicecatalog/serviceclass/storage.go]: can we also get "spec.externalID" as a selectable field?

File: pkg/registry/servicecatalog/serviceplan/strategy.go [hunk 2]
```diff
@@ -101,7 +101,7 @@ func (serviceplanRESTStrategy) PrepareForUpdate(ctx genericapirequest.Context, n
 		glog.Fatal("received a non-serviceplan object to update from")
 	}
 	// copy all fields individually?
-	newServiceplan.ServiceClassRef = oldServiceplan.ServiceClassRef
+	newServiceplan.Spec.ServiceClassRef = oldServiceplan.Spec.ServiceClassRef
 }
```
Comment (pmorie) [inline pkg/registry/servicecatalog/serviceplan/strategy.go]: Can we get the following selectable fields?

```
spec.serviceClassName
spec.externalName
spec.externalID
```

File: pkg/apis/servicecatalog/validation/serviceclass.go [hunk 3]
```diff
@@ -55,6 +55,10 @@ func validateExternalID(value string) []string {
 
 // ValidateServiceClass validates a ServiceClass and returns a list of errors.
 func ValidateServiceClass(serviceclass *sc.ServiceClass) field.ErrorList {
+	return internalValidateServiceClass(serviceclass)
+}
+
+func internalValidateServiceClass(serviceclass *sc.ServiceClass) field.ErrorList {
```
Comment (duglin) [inline pkg/apis/servicecatalog/validation/serviceclass.go]: Just curious, why was this split out from the func above?

**Comment by duglin** [comment]
So far it looks good to me

File: pkg/apis/servicecatalog/validation/serviceclass.go (hunk 3 above)
Comment (vaikas-google) [reply pkg/apis/servicecatalog/validation/serviceclass.go]: For consistency with other files in the same directory. Also there might be other validation that we might want to control differently during an update vs. create.

File: pkg/apis/servicecatalog/validation/serviceplan.go [hunk 4]
```diff
@@ -55,24 +59,30 @@ func ValidateServicePlan(serviceplan *sc.ServicePlan) field.ErrorList {
 			validateServicePlanName,
 			field.NewPath("metadata"))...)
 
-	if "" == serviceplan.ExternalID {
-		allErrs = append(allErrs, field.Required(field.NewPath("externalID"), "externalID is required"))
+	allErrs = append(allErrs, validateServicePlanSpec(&serviceplan.Spec, field.NewPath("Spec"))...)
```
[the remark this answers is not in the document]
Comment (vaikas-google) [reply pkg/apis/servicecatalog/validation/serviceplan.go]: Great, used instance.go for it and it had "Spec" :) Changed that as well here now then too...

File: pkg/apis/servicecatalog/validation/serviceplan.go [hunk 5]
```diff
@@ -55,24 +59,30 @@ func ValidateServicePlan(serviceplan *sc.ServicePlan) field.ErrorList {
...
 
-	if "" == serviceplan.ServiceClassRef.Name {
-		allErrs = append(allErrs, field.Required(field.NewPath("serviceClassRef"), "an owning serviceclass is required"))
+	if "" == spec.ServiceClassRef.Name {
+		allErrs = append(allErrs, field.Required(fldPath.Child("serviceClassRef"), "an owning serviceclass is required"))
 	}
 
-	for _, msg := range validateExternalID(serviceplan.ExternalID) {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("externalID"), serviceplan.ExternalID, msg))
+	for _, msg := range validateExternalID(spec.ExternalID) {
```
[the remark this answers is not in the document]
Comment (vaikas-google) [reply pkg/apis/servicecatalog/validation/serviceplan.go]: done.

File: pkg/apis/servicecatalog/validation/serviceclass.go [hunk 6]
```diff
@@ -64,20 +68,27 @@ func ValidateServiceClass(serviceclass *sc.ServiceClass) field.ErrorList {
 			validateServiceClassName,
 			field.NewPath("metadata"))...)
 
-	if "" == serviceclass.ServiceBrokerName {
-		allErrs = append(allErrs, field.Required(field.NewPath("brokerName"), "brokerName is required"))
+	allErrs = append(allErrs, validateServiceClassSpec(&serviceclass.Spec, field.NewPath("Spec"), true)...)
```
[the remark this answers is not in the document]
Comment (vaikas-google) [reply pkg/apis/servicecatalog/validation/serviceclass.go]: done

File: pkg/controller/controller_broker.go [hunk 7]
```diff
@@ -368,7 +368,7 @@ func (c *controller) reconcileServiceBroker(broker *v1alpha1.ServiceBroker) erro
 
 		// Delete ServiceClasses that are for THIS ServiceBroker.
 		for _, svcClass := range svcClasses {
-			if svcClass.ServiceBrokerName == broker.Name {
+			if svcClass.Spec.ServiceBrokerName == broker.Name {
```
[the remark this answers is not in the document]
Comment (vaikas-google) [reply pkg/controller/controller_broker.go]: todo? Added...

File: pkg/controller/controller_test.go [hunk 8]
```diff
@@ -366,38 +366,50 @@ func getTestServiceBrokerWithAuth(authInfo *v1alpha1.ServiceBrokerAuthInfo) *v1a
...
 	return &v1alpha1.ServicePlan{
 		ObjectMeta: metav1.ObjectMeta{Name: testServicePlanName},
-		ExternalID: planGUID,
-		Bindable:   truePtr(),
-		// ref to serviceclass
+		Spec: v1alpha1.ServicePlanSpec{
+			ExternalID:   planGUID,
+			ExternalName: testServicePlanName,
+			Bindable:     truePtr(),
+			// ref to serviceclass
```
[the remark this answers is not in the document]
Comment (vaikas-google) [reply pkg/controller/controller_test.go]: removed

**Comment by vaikas-google** [comment]
Thanks for the review, PTAL

File: pkg/registry/servicecatalog/serviceplan/storage.go [hunk 9]
```diff
@@ -89,8 +89,10 @@ func Match(label labels.Selector, field fields.Selector) storage.SelectionPredic
 
 // toSelectableFields returns a field set that represents the object for matching purposes.
 func toSelectableFields(servicePlan *servicecatalog.ServicePlan) fields.Set {
-	objectMetaFieldsSet := generic.ObjectMetaFieldsSet(&servicePlan.ObjectMeta, true)
-	return generic.MergeFieldsSets(objectMetaFieldsSet, nil)
+	spSpecificFieldsSet := make(fields.Set, 2)
+	spSpecificFieldsSet["spec.externalName"] = servicePlan.Spec.ExternalName
```
Comment (pmorie) [inline pkg/registry/servicecatalog/serviceplan/storage.go]: can we also get "spec.serviceClassRef.name" as a selectable field?

**Comment by pmorie** [comment]
My feedbacks can be addressed in a follow-up, let's get this in and lessen the rebase burden for others.

**Comment by duglin** [comment]
LGTM

**Pull request merged**